src/build-time-render/BuildTimeRender.ts

```diff
@@ -22,162 +32,205 @@ export interface BuildTimeRenderArguments {
+export default class BuildTimeRender {
 	private _paths: any[];
-	private _disabled = false;
-	private _root: string;
-	private _entries: string[];
 	private _useManifest: boolean;
-	private _hasPaths: boolean;
-	private _btrRoot: string;
+	private _manifest: any;
+	private _entries: string[];
+	private _root: string;
+	private _output?: string;
+	private _cssFiles: string[] = [];
+	private _useHistory = false;
+	private _inlinedCssClassNames: string[] = [];
+	private _head: string;
+	private _puppeteerOptions: any;
 
 	constructor(args: BuildTimeRenderArguments) {
-		const { paths = [], root = '', useManifest = false, entries } = args;
-		this._paths = ['', ...paths];
-		this._hasPaths = paths.length > 0;
+		const { paths = [], root = '', useManifest = false, entries, useHistory, puppeteerOptions } = args;
+		const path = paths[0];
+		const initialPath = typeof path === 'object' ? path.path : path;
+
+		this._puppeteerOptions = puppeteerOptions;
+		this._paths = paths;
 		this._root = root;
-		this._entries = entries;
 		this._useManifest = useManifest;
-		if (root === '') {
-			this._disabled = true;
-		}
+		this._entries = entries.map((entry) => `${entry.replace('.js', '')}.js`);
+		this._manifest = this._entries.reduce(
+			(manifest, entry) => {
+				manifest[entry] = entry;
+				return manifest;
+			},
+			{} as any
+		);
+		this._useHistory = useHistory !== undefined ? useHistory : paths.length > 0 && !/^#.*/.test(initialPath);
 	}
 
-	private _render(compiler: Compiler, location: string, htmlContent: string, root: string, output: string) {
-		const window: Window = new JSDOM(htmlContent, { runScripts: 'outside-only', pretendToBeVisual: true }).window;
-		const document: Document = window.document;
-		const parent = document.getElementById(root)!;
-		if (!this._btrRoot) {
-			this._btrRoot = parent.outerHTML;
+	private _writeIndexHtml({ html, script, path = '', styles }: RenderResult) {
+		path = typeof path === 'object' ? path.path : path;
+		const prefix = getPrefix(path);
+		if (this._head) {
+			html = html.replace(/<head>([\s\S]*?)<\/head>/gm, this._head);
 		}
-		let manifest: any = {};
-		if (this._useManifest) {
-			manifest = JSON.parse(readFileSync(path.join(output, 'manifest.json'), 'utf-8'));
-		}
-		window.eval(`
-window.location.hash = '${location}';
-window.DojoHasEnvironment = { staticFeatures: { 'build-time-render': true } };`);
-
-		this._entries.forEach((entry) => {
-			entry = this._useManifest ? manifest[`${entry}.js`] : `${entry}.js`;
-			const entryContent = readFileSync(path.join(output, entry), 'utf-8');
-			window.eval(entryContent);
-		});
+		const css = this._cssFiles.reduce((css, entry: string | undefined) => {
+			html = html.replace(`<link href="${entry}" rel="stylesheet">`, `<style>${styles}</style>`);
+			css = `${css}<link rel="stylesheet" href="${prefix}${entry}" media="none" onload="if(media!='all')media='all'" />`;
+			return css;
+		}, '');
+
+		html = html.replace(/^(\s*)(\r\n?|\n)/gm, '').trim();
```

Comment:
The following would better minify the html

```javascript
html = html.replace(/(^\s*|\r\n|\n|\t|\s$)/gm, '').trim();
```

JS Fiddle: https://jsfiddle.net/vqn3a6g5/19


Also should this occur after the `script` and `css` variables are added in? Since you don't currently the following occurs: https://gist.github.com/KaneFreeman/7cb2991eaa04e516e23e7eb1d720b04d